apps/05Mastermind.js

```diff
@@ -6,68 +6,169 @@ var prompt = require('prompt');
 prompt.start();
 
 var board = [];
+var hint = [];
 var solution = '';
 var letters = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'];
+var solutionLength = solution.length;
+
+var numberOfGuesses = 0;
 
 function printBoard() {
     for (var i = 0; i < board.length; i++) {
-        console.log(board[i])
+        console.log(board[i]);
     }
+    return board;
+}
+
+function printNumberOfGuesses() {
+    console.log('# of Guesses: ' + numberOfGuesses);
 }
 
 function generateSolution() {
     for (var i = 0; i < 4; i++) {
         var randomIndex = getRandomInt(0, letters.length);
-        solution += letters[randomIndex];
+        solution += letters[randomIndex]; //same an concatenating strings, assumes empty starting string
     }
 }
 
 function getRandomInt(min, max) {
-  return Math.floor(Math.random() * (max - min)) + min;
+    return Math.floor(Math.random() * (max - min)) + min;
 }
 
-function generateHint() {
-    // your code here
+function generateHint(guess, solution) {
+
+    console.log('solution is ' + solution);
+
+    var guessArray = guess.split('');
+    var solutionArray = solution.split('');
+
+    //1 - first check to see if any letters also match the exact position of the solution
+    //2 - then check to see if any letters of the guess are in the solution
+    var guessLength = guessArray.length;
+    var solutionLength = solutionArray.length;
+    var correctLetterLocations = 0;
+
+    for (var i = 0; i <= (solutionLength - 1); i++) {
+
+        if (guessArray[i] === solutionArray[i]) {
+            correctLetterLocations++;
+            guessArray[i] = null;
+            solutionArray[i] = null;
+        }
+
+    }
+
+    var correctLetters = 0;
+
+    //now check remaining letters to look for similar values
+    for (var j = 0; j <= (guessLength - 1); j++) {
+
+        var letterIndex = solutionArray.indexOf(guessArray[j]);
+
+        if (letterIndex > -1) {
+            correctLetters++;
+        }
+
+    }
+
+    //we don't want to double count the correct letter positions, so we subtract that number out
+    correctLetters = correctLetters - correctLetterLocations;
```

Comment:
Interesting! Another way to exclude double-counts would be to add some more logic above on line 68. Can you think of what that might be? Why are we getting double-counts even though we're nulling out the guessArray elements?